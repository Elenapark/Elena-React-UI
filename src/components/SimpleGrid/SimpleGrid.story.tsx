import React, { FC } from 'react';
import { Meta } from '@storybook/react';
import { SimpleGrid, CustomGridProps } from './SimpleGrid';

export default {
  title: 'SimpleGrid',
  component: SimpleGrid,
} as Meta<CustomGridProps>;

const style = {
  background: 'dodgerBlue',
  color: 'white',
  padding: '20px',
};

export const Default: FC = () => {
  return (
    <SimpleGrid column={3} spacingX={20} spacingY={10}>
      {Array(6)
        .fill(1)
        .map((_, idx) => {
          return <div style={{ ...style }}>{`grid-item${idx}`}</div>;
        })}
    </SimpleGrid>
  );
};
