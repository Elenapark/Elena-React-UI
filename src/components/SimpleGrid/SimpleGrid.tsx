import React, { isValidElement } from 'react';
import styled from 'styled-components';
import {
  color,
  ColorProps,
  flexbox,
  FlexboxProps,
  layout,
  LayoutProps,
  space,
  SpaceProps,
  grid,
  GridProps,
} from 'styled-system';

export interface SimpleGridProps {
  column: number;
  spacingX?: number | string;
  spacingY?: number | string;
  children?: React.ReactNode;
  [x: string]: any;
}

export type CustomGridProps = SimpleGridProps & SpaceProps & LayoutProps & ColorProps & FlexboxProps & GridProps;

const SimpleGrid: FCC<CustomGridProps> = ({ column = 2, spacingX = 10, spacingY = 10, children, ...props }) => {
  const refinedChildren = React.Children.map(children, (child, idx) => {
    if (isValidElement(child)) {
      const clonedChild = React.cloneElement(child, {
        ...child.props,
        key: `SimpleGrid-item${idx}`,
        'data-testid': 'simple-grid-item-comp',
      });

      return clonedChild;
    } else {
      return (
        <div key={`SimpleGrid-item${idx}`} data-testid="simple-grid-item-comp">
          {child}
        </div>
      );
    }
  });

  return (
    <StyledSimpleGrid data-testid="simple-grid-comp" column={column} spacingX={spacingX} spacingY={spacingY} {...props}>
      {refinedChildren}
    </StyledSimpleGrid>
  );
};

export default SimpleGrid;

const StyledSimpleGrid = styled.div<SimpleGridProps>`
  ${color}
  ${flexbox}
	${layout}
	${space}
	${grid}

  display: grid;
  grid-template-columns: ${(props) => `repeat(${props.column},1fr)`};
  grid-column-gap: ${(props) => `${props.spacingX}px`};
  grid-row-gap: ${(props) => `${props.spacingY}px`};
`;
