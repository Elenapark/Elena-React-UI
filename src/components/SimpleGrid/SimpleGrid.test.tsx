import React from 'react';
import { SimpleGrid } from './SimpleGrid';
import { render, screen } from '@testing-library/react';

const children = [<div>그리드1</div>, <div>그리드2</div>];

const defaultProps = {
  column: 2,
  spacingX: 10,
  spacingY: 10,
  children,
  className: 'simple-grid',
};

describe('<SimpleGrid />', () => {
  it('should render', () => {
    render(<SimpleGrid {...defaultProps} />);
    expect(screen.getByTestId('simple-grid-comp')).toBeInTheDocument();
  });

  it('should render children', () => {
    render(<SimpleGrid {...defaultProps} />);
    expect(screen.getAllByTestId('simple-grid-child-comp')).toHaveLength(2);
  });

  it('has grid styling', () => {
    render(<SimpleGrid {...defaultProps} />);
    expect(screen.getByTestId('simple-grid-comp')).toHaveStyle(`display: grid`);
    expect(screen.getByTestId('simple-grid-comp')).toHaveStyle(`grid-template-columns: repeat(2,1fr)`);
  });
});
