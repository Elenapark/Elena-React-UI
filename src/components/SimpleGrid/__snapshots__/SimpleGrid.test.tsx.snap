// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SimpleGrid /> should render 1`] = `
<div
  class="sc-bcXHqe hMRnOh"
  data-testid="simple-grid-comp"
>
  <div
    data-testid="simple-grid-item-comp"
  >
    그리드1
  </div>
  <div
    data-testid="simple-grid-item-comp"
  >
    그리드2
  </div>
</div>
`;
